andela/ah-robotics
# #162949072 An Email should be sent to a user upon successful registration.

#### What does this PR do?
Sends an activation email to the user upon successful registration. Once clicked, the user account is activated.

#### Description of Task to be completed?

- Send an email to the user on registration
- Verify user account on clicking `verify account`

#### How should this be manually tested?
1. Clone this repository on your local machine using `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using `cd ah-robotics`
3. Change to the specific branch `ft-user-verification-email-162949072` using `git checkout ft-user-verification-email-162949072`
4. Create a virtual environment using the following command: `python3 -m virtualenv env`
5. Activate the virtual environment using `source env/bin/activate`
6. Install all the requirements using `pip install -r requirements.txt`

### Database setup

Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running psql, follow these steps to create a new user and a database which will be used in the application.

  1. Start the postgres database server using the command:
  
      `pg_ctl -D /usr/local/var/postgres start`

  2. Run psql interactive terminal

      `psql postgres`

  3. Create a new user:

      `CREATE USER sample_username WITH PASSWORD 'sample_password';`

  4. Grant privileges to the user:

      `ALTER USER sample_username CREATEDB;`

  5. Create a database:

      `CREATE DATABASE sample_database_name WITH OWNER sample_username;`

  6. Exit psql interactive with the following command: `\q`

  7. Create a **.env** file and add the following configuration variables

```source-json
      source env/bin/activate
      export DB_NAME="database_name"
      export DB_USER="sample_username"
      export DB_PASS="sample_password"
      export DB_HOST="host_name_eg_localhost"
      export DB_PORT="port_number_eg_5432"
      export TIME_DELTA="30"
      export EMAIL_HOST="smtp.gmail.com"
```
 Refer to the email pinned on the Team Slack channel and add them to the .env file
```
      export EMAIL_HOST_USER="user@example.com"
      export EMAIL_HOST_PASSWORD="email_password"
```
Finally, add the email port to the .env file which is to remain as indicated below.
```
      export EMAIL_PORT=587
```

  7. Source .env to set the configuration variables in the terminal:

      `source .env`
  8. Run migrations through the terminal to create the tables in the database:
      (i) Make the migrations using: `python manage.py makemigrations`.
      (ii) Apply the migrations using: `python manage.py migrate`

9. Collect static files

Run the following command to collect static files.

`python manage.py collectstatic`
      
### Testing the features

1. Run the server from the terminal using `python manage.py runserver`.
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to test.
3. On the body section on postman select _raw_ and type as _JSON/application_.
4. Add your payload to the **body** section using the following format:

```
    {
        "user": {
            "username": "sample_username",
            "email": "sample_email",
            "password": "sample_password"
            }
    }
```

4. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
5. Open your browser, go to the inbox of the email you inserted above, open the verification email and click the button labelled `click here` to confirm registration.
6. To test the `resend verification message` link, go to [Postman](https://www.getpostman.com/) again on your local machine, and Paste the URL `http://127.0.0.1:8000/api/v1/users/resend-email/` to test.
7. On the body section on postman select _raw_ and type as _JSON/application_.
8. Add your payload to the **body** section using the following format:

```
    {
        "user": {
            "email": "sample_email"
            }
    }
```
9. Once all the fields are inserted, ensure that it is a **POST** request in postman and click the **Send** button.
10. If the verification link has been resent successfully, recheck the inbox of the email you inserted above, go to the resend verification email and click the button labelled `click here` to confirm registration.

### Running tests
1. Go to the terminal and ensure that the server is not running.
2. Run the tests by typing in the following command: `python manage.py test`

#### Any background context you want to provide?
This is building upon the registration feature. Once the user is registered, a token is generated which is appended to the verification URL. When the user clicks `verify`, the token appended is checked and the account verified.

#### What are the relevant pivotal tracker stories?
[#162949072](https://www.pivotaltracker.com/n/projects/2235171/stories/162949072)

#### Screenshots (if appropriate)
**Successful signup**
<img width="1004" alt="screen shot 2019-01-22 at 18 00 32" src="https://user-images.githubusercontent.com/14871749/51543910-bc3f9f00-1e6f-11e9-8bfa-e1ea4a3e0012.png">

**Verification Email**
![image](https://user-images.githubusercontent.com/10106044/51597752-0a56b000-1f0d-11e9-8fb2-20854bca3a7d.png)

**Resend Verification Email**
![image](https://user-images.githubusercontent.com/10106044/51597983-9cf74f00-1f0d-11e9-912c-f6e52434c63a.png)






**Verification success message**
![image](https://user-images.githubusercontent.com/10106044/51489447-64466100-1db9-11e9-8bf1-0c78a4f5e700.png)

**Resend verification link successfully**
![image](https://user-images.githubusercontent.com/10106044/51563017-456ccb00-1e9c-11e9-9973-fd965f5ad1ad.png)

**Invalid email on resend email endpoint**
![image](https://user-images.githubusercontent.com/10106044/51563129-9a104600-1e9c-11e9-8f17-d5842220298b.png)

**Blank email field when trying to resend verification link**
![image](https://user-images.githubusercontent.com/10106044/51563176-b6ac7e00-1e9c-11e9-87e8-e01b3ffd87b8.png)

**User already verified error message on resend email endpoint**
![image](https://user-images.githubusercontent.com/10106044/51563214-cc21a800-1e9c-11e9-948a-b4d0d8d5610f.png)

**Non existent user email error when trying to resend verification link**
![image](https://user-images.githubusercontent.com/10106044/51563439-7c8fac00-1e9d-11e9-99b0-694022afec98.png)


## Questions:
Questions are welcome
